vars/secretScanner.groovy

```diff
@@ -0,0 +1,165 @@
+// private
+@NonCPS // Don't run this in the Jenkins sandbox so that use (groovy.time.TimeCategory) will work
+def getCommitCheckDate(scanWindowHrs) {
+  use (groovy.time.TimeCategory) {
+    return (new Date() - scanWindowHrs.hours - 10.minutes).format("yyyy-MM-dd'T'HH:mm:ss'Z'", TimeZone.getTimeZone('UTC'))
+  }
+}
+
+// private
+def runTruffleHog(dockerImgName, repoName, excludeStrings, commitShas=null) {
+  // truffleHog seems to alway exit with code 1 even though it appears to run fine
+  // which fails the build so we need the || true to ignore the exit code and carry on
+  def truffleHogCmd = "docker run $dockerImgName --json https://github.com/${repoName}.git || true"
+  def truffleHogResults = sh returnStdout: true, script: truffleHogCmd
+  def secretMessages = []
+
+  truffleHogResults.trim().split('\n').each {
+    if (it.length() == 0) return  // readJSON won't accept an empty string
+    def result = readJSON text: it
+
+    if (!commitShas || commitShas.contains(result.commitHash)) {
+      if (!result.stringsFound.every { it in excludeStrings }) {
+        def message = "Reason: $result.reason\n" +
+                      "Date: $result.date\n" +
+                      "Repo: $repoName\n" +
+                      "Branch: $result.branch\n" +
+                      "Hash: $result.commitHash\n" +
+                      "File path: $result.path\n" +
+                      "Strings found: $result.stringsFound\n"
+
+        secretMessages.add(message)
+      }
+    }
+  }
+
+  return secretMessages
+}
+
+// private
+def getNumPages(curlHeaders) {
+  def linkHeader = curlHeaders.split('\n').find { header -> header.toLowerCase().startsWith('link:') }
+
+  if (linkHeader) {
+    def lastLink = linkHeader.split(',').find { link -> link.contains('rel="last"') }
+
+    if (lastLink) {
+      return lastLink[(lastLink.lastIndexOf('&page=')+6)..(lastLink.lastIndexOf('>')-1)]
+    }
+  }
+
+  return "1"
+}
+
+// private
+def getMatchingRepos(curlAuth, githubOwner, repositoryPrefix) {
+  def githubReposUrl = "https://api.github.com/users/$githubOwner/repos?per_page=100"
+
+  def curlHeaders = sh returnStdout: true, script: "$curlAuth --head $githubReposUrl"
+  def numPages = getNumPages(curlHeaders)
+
+  echo "Number of pages of repos: $numPages"
+
+  def matchingRepos = []
+  def matchStr = "$githubOwner/$repositoryPrefix".toLowerCase()
+
+  (numPages as Integer).times { page ->
+    def reposResults = sh returnStdout: true, script: "$curlAuth $githubReposUrl\\&page=${page+1}"
+    def allRepos = readJSON text: reposResults
+
+    allRepos.each { repo ->
+      if (repo.full_name.toLowerCase().startsWith(matchStr)) {
+        matchingRepos.add(repo.full_name)
+      }
+    }
+  }
+
+  echo "Matching repos: $matchingRepos"
+
+  return matchingRepos
+}
+
+// private
+def reportSecrets(secretMessages, repo, channel) {
```

Comment:
Would this fit better in the `notifySlack` file?

Reply:
Potentially the bit for the sending message yes.  There is a bit of conditional logic in here to switch reporting on and off (based on an empty channel string).

What I would suggest then is maybe a generic a function in notifySlack parameterised to send a message to a channel?

Reply:
Makes sense